### PR TITLE
Building Interaction Real-Time Sync

### PR DESCRIPTION
**Pull request by Baldie-dev**
This is part of the #104 

Following Laboratory events were synchronized between players on the same planet:

- Reset recipe
- Select a new cube recipe
- Change laboratory to research mode
- Put items for cube production
- Withdraw items for cube production
- Withdraw produced cubes
- Put cubes for research
- Withdraw cubes for research

+Note: Small fix for the planet broadcast, packet will not be sent back to the player who sent it.

Video on [youtube](https://www.youtube.com/watch?v=aExI_Krt1To).

### CODE REVIEW

**Comment by Baldie-dev** [comment]
Following were added:

- Sync insertion / withdrawing sail bullets from Railguns
- Change orbit of Railguns
- Sync insertion / withdrawing sail rockets from Silo Launchers
- Sync changing mode of Ray Reciever
- Synch insertion/withdrawing of the gravitational lens from Ray Reciever

+ Note: Small fix for the loading server's savefile.

Video on [youtube](https://www.youtube.com/watch?v=S3fMChEgp1Q).

File: NebulaHost/PacketProcessors/Factory/Laboratory/LaboratoryUpdateEventProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                    pool[packet.LabIndex].produced[0] = 0;
+                }
+                else if (packet.ProductId == -2)
+                {
+                    //Research recipe reseted
+                    pool[packet.LabIndex].SetFunction(false, 0, 0, GameMain.localPlanet.factory.entitySignPool);
+                }
+                else if (packet.ProductId == -1)
+                {
+                    //Center chenged to research-mode
```
Comment (hubastard) [inline NebulaHost/PacketProcessors/Factory/Laboratory/LaboratoryUpdateEventProcessor.cs:28]: Small typo *changed

**Review by hubastard — APPROVED** [review]

File: NebulaHost/PacketProcessors/Factory/Laboratory/LaboratoryUpdateEventProcessor.cs (hunk 1 above)
Comment (Baldie-dev) [reply NebulaHost/PacketProcessors/Factory/Laboratory/LaboratoryUpdateEventProcessor.cs]: fixed :)

**Comment by Baldie-dev** [comment]
The following has been added:

- Adding, changing, and removing fuel from the Power Generators.
- Setting priorities and filters in the splitter.
- Setting and resetting filters in the inserter.

**Comment by Baldie-dev** [comment]
I will add a few more synchronization events and then we can merge :)

**Comment by Baldie-dev** [comment]
The following has been added:

- Withdrawing items from the miners (miner, water pump, oil extractor)
- Depositing and withdrawing fluids from fluid storages
- Changing output/input switch in fluid storage
- Changing modes of Power exchanger
- Inserting and withdrawing accumulators from power exchanger

**Review by hubastard — APPROVED** [review]

**Pull request merged**